components/layout/display_list/builder.rs

```diff
@@ -840,33 +840,19 @@ impl FragmentDisplayListBuilding for Fragment {
         // inefficient. What we really want is something like "nearest ancestor element that
         // doesn't have a fragment".
 
-        // 'background-clip' determines the area within which the background is painted.
-        // http://dev.w3.org/csswg/css-backgrounds-3/#the-background-clip
-        let mut bounds = *absolute_bounds;
-
         // Quote from CSS Backgrounds and Borders Module Level 3:
         //
         // > The background color is clipped according to the background-clip value associated
         // > with the bottom-most background image layer.
         let last_background_image_index = background.background_image.0.len() - 1;
-        let color_clip = get_cyclic(&background.background_clip.0, last_background_image_index);
-
-        // Adjust the clipping region as necessary to account for `border-radius`.
-        let mut border_radii = build_border_radius(absolute_bounds, style.get_border());
-
-        match color_clip {
-            BackgroundClip::BorderBox => {},
-            BackgroundClip::PaddingBox => {
-                let border = style.logical_border_width().to_physical(style.writing_mode);
-                bounds = bounds.inner_rect(border);
-                border_radii = calculate_inner_border_radii(border_radii, border);
-            },
-            BackgroundClip::ContentBox => {
-                let border_padding = self.border_padding.to_physical(style.writing_mode);
-                bounds = bounds.inner_rect(border_padding);
-                border_radii = calculate_inner_border_radii(border_radii, border_padding);
-            },
-        }
+        let color_clip = *get_cyclic(&background.background_clip.0, last_background_image_index);
+        let (bounds, border_radii) = compute_background_clip(
+            color_clip,
+            *absolute_bounds,
+            style.logical_border_width().to_physical(style.writing_mode),
```

Comment:
As a followup, let's double-check that computing the background-clip in terms of the padding of a fragment, but the border of another one is the right thing to do. Looks fishy.

Reply:
> Looks fishy.

It probably is. For block fragments it does not make a difference as it is always the same fragment but for inlines it is probably wrong. However I don't know how to correctly compute border_padding for inlines. I assumed it was padding + border for all four sides but [there is more going on](https://github.com/pyfisch/servo/blob/0ddcf33c97c1a1ca2ec26f5e453fc327b35a2f59/components/layout/fragment.rs#L1272-L1313).